service/src/main/java/org/cbioportal/service/impl/MutationalSignatureServiceImpl.java

```diff
@@ -0,0 +1,131 @@
+package org.cbioportal.service.impl;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedList;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import org.cbioportal.model.GeneticProfile;
+import org.cbioportal.model.MutationalSignature;
+import org.cbioportal.model.SNPCount;
+import org.cbioportal.model.Sample;
+import org.cbioportal.persistence.GeneticProfileRepository;
+import org.cbioportal.persistence.MutationalSignatureRepository;
+import org.cbioportal.persistence.SampleRepository;
+import org.cbioportal.service.MutationalSignatureService;
+
+@Service
+public class MutationalSignatureServiceImpl implements MutationalSignatureService {
+
+	public static class MutationalSignatureFactory {
+		private static final Map<String, String> COMPLEMENTARY_NUCLEOTIDE = new HashMap<>();
+		private static final String[] CANONICAL_SNP_TYPES = new String[]{"C>A", "C>G", "C>T", "T>A", "T>C", "T>G"};
+		private static final Map<String, Integer> CANONICAL_SNP_TYPES_INDEX;
+		
+		static {
+			COMPLEMENTARY_NUCLEOTIDE.put("A","T");
```

Comment:
sure .. let's wait. There may be some excellent third party libraries too for representing nucleotide sequences (with operations like reverse complement), amino acid sequences, codon mappings and non-synonymous mutation detection, etc. We should probably prefer an established library to crafting our own.